tests/cxx/badinc.h

```diff
@@ -297,10 +296,12 @@ typedef std::vector<I2_Class> H_I2Class_Vector_Unused;
 // IWYU: I2_Enum is...*badinc-i2.h
 typedef I2_TemplateClass<I2_Enum> H_TemplateTypedef;
 
-// IWYU: I2_Struct needs a declaration
+// IWYU: I2_Struct is...*badinc-i2.h
```

Comment:
This one looks a little weird. Why not a forward declare?

Reply:
I've decided simply to expand IWYU policy (stating that typedef author should provide underlying type by default) to typedef underlying type components. `I2_Struct` is just a component of `I2_Struct*`. It should be possible to override it by means of fwd-declaring and not including the header directly.

Reply:
Nice, I see. 

I keep going back and forth... At first I thought it felt a little blunt to consider `I2_Struct` a component of `I2_Struct*`, then I thought it makes a lot of sense for consistency.

But there seems to be tension between the definition loc and the usage loc now;

* typedef location is responsible for the full type even if it only uses a pointer
* can work around that by forward-declaring the type
* ... but that implies the usage loc is responsible

Or did I misunderstand that?

Reply:
> ... but that implies the usage loc is responsible

Of course. If a typedef author doesn't provide a component which is needed at the use site (e.g. it is absent in `provided_types` / `blocked_types`), then a user should `#include` the corresponding header.

In fact, pointer typedefs are not always handled correctly at use site both on master and my branch. IWYU often suggests to `#include` an underlying type header when it is redundant. I want to address this with something like [such a code](https://github.com/bolshakov-a/include-what-you-use/commit/a043e0013fe0b8c2262eeb32f8e958f5368f2c7f).

Reply:
Seems like the original IWYU authors meant that pointed-to types may also be subject to "providing", as the example with function at the end of ["Author intent" section](https://github.com/include-what-you-use/include-what-you-use/blob/0.20/docs/WhatIsAUse.md#author-intent) shows. And later:
> As of May 2011, the rule does not apply when returning a pointer or reference
> ...
> This is considered a bug ...

Honestly, I don't think that this "bug" regarding function return type will be fixed some day because IWYU quickly loses information that a used type is from a function return. But for typedefs, it can be done because typedef type is tracked in many cases.

`RemovePointersAndReferencesAsWritten` calls in [`GetCallerResponsibleTypesForTypedef`](https://github.com/include-what-you-use/include-what-you-use/blob/0.20/iwyu.cc#L1454) and [`VisitTypedefNameDecl`](https://github.com/include-what-you-use/include-what-you-use/blob/0.20/iwyu.cc#L1759) versions from the master branch cause as well a thought that pointed-to types could be "provided" and should obey the default policy rules.

Btw, I want to introduce some day a command line flags to suppress that policies and require forward declarations of components of typedefs, function arguments and returns.

Reply:
I have an old branch to play with custom AST traversal to be able to analyze the constituent parts of function types: https://github.com/kimgr/include-what-you-use/tree/function-type-traversal. I think that might be a way to maintain the information about whether a type is found as a function return type, exception spec, or something else.

Now if only I could get my nose over water long enough to get it more finished... :-)

Reply:
I think I agree with the more aggressive type requirement for pointers/references, but it's hard to see how it will affect real-world codebases.

Reply:
> I have an old branch to play with custom AST traversal to be able to analyze the constituent parts of function types

Interesting, thanks! I think, though, that simple setting fwd-decl-context flag is insufficient for for determination of whether a type is a provided component of a fn-return type, so I've decided to use `ASTNode` stack for that purpose:
https://github.com/bolshakov-a/include-what-you-use/commit/d4666e32c74575c8b0f5197b109c5a471e3c17d1

Btw, I've just realized that the latest refactoring in this PR doesn't fit very well to the suggested way of reporting provided components of a fn-return:
https://github.com/bolshakov-a/include-what-you-use/commit/4a28fbc8c0371d41aed6341439e55a549dbafdfb#diff-7ca107c8b836c1765e8b735b74f06833bec48f308a76017886cee1d5736002d9R4212
Maybe, it can be replaced with some scoped setting of "provided types" set which would be returned from `IwyuAstConsumer::GetBlockedTypes()`.

> how it will affect real-world codebases

If an old header follows the old IWYU rules, then it already has the fwd-declaration and doesn't have `#include` in that case, generally. So the new rules shouldn't affect it. In a new headers, writers should pay more attention on "provision" policy which they want. But for value types, it is the case even with the old rules, of course. E.g., some style guide may insist that function parameter and return types should be fwd-declared regardless of whether they are pointers or not. So, code writers should check and correct IWYU output by itself.